torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -1159,42 +1191,64 @@ class ShapePropagator {
     //   - has a bool keepdim argument
     static const register_formula_for dim_reduce_ops_with_integer_upcast{
         {
-            "aten::prod(Tensor self, int dim, bool keepdim) -> Tensor",
+            "aten::prod(Tensor self, int dim, bool keepdim, *, int? dtype) -> Tensor",
+        },
+        [](Node* node) -> type_vec_t {
+          auto maybe_keepdim = node->get<bool>(attr::keepdim);
+          at::optional<IValue> opt_dtype = node->get(attr::dtype);
+          return reduce_op_handler(
+              node, /*num_reduce_dim=*/*maybe_keepdim?0:1, /*integer_upcast=*/true, opt_dtype);
+        }};
+
+    // Requirements:
+    //   dims           : preserved
+    //   scalar type    : preserved unless specified.
```

Comment:
`scalar type    : preserved unless specified` is inaccurate since integer_upcast is true

Reply:
right, fixed